deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/graph/ComputationGraph.java

```diff
@@ -1880,7 +1939,15 @@ protected void calcBackpropGradients(boolean truncatedBPTT, INDArray... external
                         continue;
                     }
 
-                    epsilons[x] = epsilons[x].leverageTo(workspaceExternal);
+                    if(wsm == WorkspaceMode.SEPARATE) {
+                        epsilons[x] = epsilons[x].leverageOrDetach(workspaceExternal);
+                    } else if(wsm == WorkspaceMode.SINGLE){
+                        //TODO we can't simply leverage to workspaceExternal here:
+                        //If using SINGLE mode, workspaceExternal *is* active but then leverageTo becomes a no-op
+                        //and hence we aren't actually moving it out of the current workspace. Consequently, the
+                        //epsilons will be invalidated at the end of the current vertex for loop
+                        epsilons[x] = epsilons[x].detach();
```

Comment:
i'm still not sure why you're detaching it here. In case of SINGLE workspace mode, memory within survives full iteration over 1 dataset